### PR TITLE
feat(connector): add authorize, capture, void, psync, refund, rsync for Forte connector

### PR DESCRIPTION
**Pull request by prasunna09**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Updating support for [Forte](https://www.forte.net) payment gateway
Payment Method: card
Supported payment flows:

Authorize
Capture
Void
PSync
Refund
RSync

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Screenshot 2023-04-26 at 6 18 26 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [x] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/forte/transformers.rs [hunk 1]
```diff
@@ -2,90 +2,245 @@ use masking::Secret;
...
+    },
     core::errors,
-    types::{self, api, storage::enums},
+    pii::{self},
+    types::{self, api, storage::enums, transformers::ForeignFrom},
 };
 
-#[derive(Default, Debug, Serialize, Eq, PartialEq)]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "snake_case")]
```
Comment (Abhicodes-crypto) [inline crates/router/src/connector/forte/transformers.rs:14]: Since the fields are already in snake case. I guess this line isn't necessary.

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/forte/transformers.rs [hunk 2]
```diff
@@ -2,90 +2,241 @@ use masking::Secret;
...
-#[derive(Default, Debug, Serialize, Eq, PartialEq)]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
 pub struct FortePaymentsRequest {
-    amount: i64,
-    card: ForteCard,
+    action: String,
+    authorization_amount: f64,
+    subtotal_amount: f64,
+    billing_address: BillingAddress,
+    card: Card,
```
[the remark this answers is not in the document]
Comment (prasunna09) [reply crates/router/src/connector/forte/transformers.rs:19]: Yes, I have mailed them, waiting for their reply

File: crates/router/src/connector/forte/transformers.rs [hunk 3]
```diff
@@ -2,90 +2,236 @@ use masking::Secret;
...
+
+impl ForeignFrom<(ForteResponseCode, ForteAction)> for enums::AttemptStatus {
+    fn foreign_from(item: (ForteResponseCode, ForteAction)) -> Self {
+        match item.0 {
+            ForteResponseCode::A01 => match item.1 {
+                ForteAction::Authorize => Self::Authorized,
+                ForteAction::Sale => Self::Pending,
+            },
+            ForteResponseCode::A05 | ForteResponseCode::A06 => Self::Authorizing,
+            _ => Self::Failure,
```
Comment (jagan-jaya) [inline crates/router/src/connector/forte/transformers.rs:166]:
```suggestion
    fn foreign_from((response_code, action): (ForteResponseCode, ForteAction)) -> Self {
        match response_code {
            ForteResponseCode::A01 => match action {
                ForteAction::Authorize => Self::Authorized,
                ForteAction::Sale => Self::Pending,
            },
            ForteResponseCode::A05 | ForteResponseCode::A06 => Self::Authorizing,
            _ => Self::Failure,
```

File: crates/router/src/connector/forte/transformers.rs [hunk 4]
```diff
@@ -2,90 +2,236 @@ use masking::Secret;
...
         }
     }
 }
 
-#[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[derive(Default, Debug, Clone, Deserialize, PartialEq)]
+pub struct CardResponse {
+    name_on_card: String,
```
Comment (jagan-jaya) [inline crates/router/src/connector/forte/transformers.rs:173]: mask it, name is PII data

File: crates/router/tests/connectors/forte.rs [hunk 5]
```diff
@@ -56,35 +85,55 @@ async fn should_only_authorize_payment() {
...
+            Some(types::PaymentsCaptureData {
+                connector_meta,
+                ..utils::PaymentCaptureType::default().0
+            }),
+            get_default_payment_info(),
+        )
         .await
         .expect("Capture payment response");
-    assert_eq!(response.status, enums::AttemptStatus::Charged);
+    assert_eq!(response.status, enums::AttemptStatus::Pending);
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/forte.rs:105]: add reason for why it is pending, when this will succeed

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/forte/transformers.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/router/src/connector/forte/transformers.rs:18]: create a GitHub issue and add in the code

File: crates/router/src/connector/forte/transformers.rs [hunk 6]
```diff
@@ -2,90 +2,241 @@ use masking::Secret;
...
-            _ => Err(errors::ConnectorError::FailedToObtainAuthType.into()),
+            _ => Err(errors::ConnectorError::FailedToObtainAuthType)?,
         }
     }
 }
 // PaymentsResponse
-#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
-#[serde(rename_all = "lowercase")]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "snake_case")]
```
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/router/src/connector/forte/transformers.rs]: change it to lowercase

File: crates/router/src/connector/forte/transformers.rs [hunk 7]
```diff
@@ -2,90 +2,234 @@ use masking::Secret;
...
 
-#[derive(Default, Debug, Serialize, Eq, PartialEq)]
-pub struct ForteCard {
-    name: Secret<String>,
-    number: Secret<String, common_utils::pii::CardNumber>,
-    expiry_month: Secret<String>,
-    expiry_year: Secret<String>,
-    cvc: Secret<String>,
-    complete: bool,
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
```
Comment (ArjunKarthik) [inline crates/router/src/connector/forte/transformers.rs]: As discussed use only necessary things

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by HyperSwitchers — APPROVED** [review]

**Review by jagan-jaya — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

**Review by jagan-jaya — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Pull request merged**